File: packages/extractor-python/src/source-index.ts
/** Bounded stdlib-AST topology and symbol index for progressive partial Python graphs. */

import { Buffer } from "node:buffer";
import { realpathSync, statSync } from "node:fs";
import { isAbsolute, relative, resolve } from "node:path";
import {
  compareBinaryStrings,
  closePythonExtractionFiles,
  nodeIdSchema,
  type CompactGraphSymbolEntry,
  type NodeId,
  type NodeKind,
} from "@meridian/core";
export { closePythonExtractionFiles } from "@meridian/core";
import { runPythonSourceIndexer } from "./analyzer";

const MAX_FILES = 100_000;
const MAX_SOURCE_BYTES = 1024 * 1024 * 1024;
const MAX_FILE_BYTES = 16 * 1024 * 1024;
const MAX_SYMBOLS = 500_000;
const MAX_ADJACENCY = 1_000_000;
const MAX_PATH_BYTES_TOTAL = 8 * 1024 * 1024;
const MAX_SELECTED_FILES = 20_000;
const MAX_PATH_BYTES = 4_096;
const INDEXED_KINDS = new Set<NodeKind>(["function", "method", "module", "class", "interface"]);

export interface PythonSourceIndexLimits {
  maxFiles: number;
  maxSourceBytes: number;
  maxFileBytes: number;
  maxSymbols: number;
  maxAdjacencyEntries: number;
}

export interface PythonSourceTopologyFile {
  path: string;
  fileId: NodeId;
  neighbors: NodeId[];
  uncertain: boolean;
}

export interface PythonSourceIndex {
  seeds: string[];
  symbols: CompactGraphSymbolEntry[];
  topology: PythonSourceTopologyFile[];
  sourceFileCount: number;
  sourceBytes: number;
  adjacencyEntries: number;
}

export interface PythonSourceIndexOptions {
  root: string;
  seeds: readonly string[];
  exclude?: readonly string[];
  limits: PythonSourceIndexLimits;
}

export interface InitialPythonProjectionOptions {
  root: string;
  seeds: readonly string[];
  depth: number;
  exclude?: readonly string[];
  maxSelectedFiles?: number;
}

export interface InitialPythonProjectionFiles {
  /** Requested weak-adjacency BFS files; these alone define projection/frontier state. */
  files: string[];
  /** BFS files plus minimal parsed allocator stabilizers needed for exact full-source IDs. */
  extractionFiles: string[];
  seeds: string[];
  frontier: string[];
}

/** Select a complete weak-import neighbourhood without constructing the semantic Python graph. */
export async function selectInitialPythonProjectionFiles(
  options: InitialPythonProjectionOptions,
): Promise<InitialPythonProjectionFiles | null> {
  if (!Number.isSafeInteger(options.depth) || options.depth < 0 || options.depth > 8) return null;
  const maxSelectedFiles = options.maxSelectedFiles ?? MAX_SELECTED_FILES;
  if (!Number.isSafeInteger(maxSelectedFiles) || maxSelectedFiles <= 0 || maxSelectedFiles > MAX_SELECTED_FILES) {
    return null;
  }
  if (options.seeds.length === 0) {
    return existingDirectory(options.root)
      ? { files: [], extractionFiles: [], seeds: [], frontier: [] }
      : null;
  }
  let indexed: PythonSourceIndex;
  try {
    indexed = await scanPythonWorkspaceSources({
      root: options.root,
      seeds: options.seeds,
      exclude: options.exclude,
      limits: defaultLimits(),
    }, false);
  } catch {
    return null;
  }
  const selected = bfs(indexed.topology, indexed.seeds, options.depth, maxSelectedFiles);
  if (selected === null) return null;
  const extractionFiles = closePythonExtractionFiles(indexed.topology, selected.files, maxSelectedFiles);
  return extractionFiles === null ? null : { ...selected, extractionFiles };
}

/** Build the complete source-backed Python symbol inventory and its shared import topology. */
export async function indexPythonWorkspaceSources(
  options: PythonSourceIndexOptions,
): Promise<PythonSourceIndex> {
  return scanPythonWorkspaceSources(options, true);
}

async function scanPythonWorkspaceSources(
  options: PythonSourceIndexOptions,
  includeSymbols: boolean,
): Promise<PythonSourceIndex> {
  validateLimits(options.limits);
  const root = canonicalDirectory(options.root);
  const seeds = normalizeSeeds(root, options.seeds);
  const raw = await runPythonSourceIndexer(root, {
    seeds,
    exclude: [...(options.exclude ?? [])],
    includeSymbols,
    ...options.limits,
    maxPathBytes: MAX_PATH_BYTES_TOTAL,
  });
  return parseIndex(raw, seeds, options.limits, includeSymbols);
}

function parseIndex(
  raw: unknown,
  expectedSeeds: readonly string[],
  limits: PythonSourceIndexLimits,
  includeSymbols: boolean,
): PythonSourceIndex {
  if (!isRecord(raw) || !exactKeys(raw, [
    "adjacencyEntries",
    "files",
    "seeds",
    "sourceBytes",
    "sourceFileCount",
    "symbols",
  ]) || !Array.isArray(raw.files) || !Array.isArray(raw.symbols) || !Array.isArray(raw.seeds)) {
    throw new TypeError("Python source index is invalid");
  }
  const sourceFileCount = boundedInteger(raw.sourceFileCount, 0, limits.maxFiles);
  const sourceBytes = boundedInteger(raw.sourceBytes, 0, limits.maxSourceBytes);
  const adjacencyEntries = boundedInteger(raw.adjacencyEntries, 0, limits.maxAdjacencyEntries);
  if (raw.files.length !== sourceFileCount || raw.symbols.length > limits.maxSymbols) {
    throw new TypeError("Python source index is invalid");
  }

  const topology: PythonSourceTopologyFile[] = [];
  const byId = new Map<string, PythonSourceTopologyFile>();
  const byPath = new Map<string, PythonSourceTopologyFile>();
  let observedAdjacency = 0;
  for (const value of raw.files) {
    if (!isRecord(value) || !exactKeys(value, ["fileId", "neighbors", "path", "uncertain"])
      || !safePath(value.path) || typeof value.fileId !== "string"
      || !pythonFileId(value.fileId) || value.uncertain !== false || !Array.isArray(value.neighbors)
      || byPath.has(value.path) || byId.has(value.fileId)) {
      throw new TypeError("Python source topology is invalid");
    }
    const neighbors: NodeId[] = [];
    const neighborIds = new Set<NodeId>();
    for (const neighbor of value.neighbors) {
      if (typeof neighbor !== "string" || !pythonFileId(neighbor) || neighbor === value.fileId
        || neighborIds.has(neighbor)) {
        throw new TypeError("Python source topology is invalid");
      }
      neighborIds.add(neighbor);
      neighbors.push(neighbor);
      observedAdjacency += 1;
    }
    neighbors.sort(compareBinaryStrings);
    const file = { path: value.path, fileId: value.fileId, neighbors, uncertain: false } satisfies PythonSourceTopologyFile;
    topology.push(file);
    byPath.set(file.path, file);
    byId.set(file.fileId, file);
  }
  topology.sort((left, right) => compareBinaryStrings(left.path, right.path));
  if (observedAdjacency !== adjacencyEntries) throw new TypeError("Python source topology is invalid");
  for (const file of topology) {
    for (const neighborId of file.neighbors) {
      const neighbor = byId.get(neighborId);
      if (neighbor === undefined || !neighbor.neighbors.includes(file.fileId)) {
        throw new TypeError("Python source topology is invalid");
      }
    }
  }

  const seeds = sortedPaths(raw.seeds, byPath);
  if (includeSymbols) {
    if (!sameStrings(seeds, expectedSeeds)) throw new TypeError("Python source-index seeds changed");
  } else {
    const requested = new Set(expectedSeeds);
    if (seeds.some((seed) => !requested.has(seed))) {
      throw new TypeError("Python source-index seeds changed");
    }
  }
  const symbols = includeSymbols ? raw.symbols.map((value) => parseSymbol(value, byId)) : [];
  if (!includeSymbols && raw.symbols.length !== 0) throw new TypeError("Python topology scan emitted symbols");
  return { seeds, symbols, topology, sourceFileCount, sourceBytes, adjacencyEntries };
}

function parseSymbol(value: unknown, byId: ReadonlyMap<string, PythonSourceTopologyFile>): CompactGraphSymbolEntry {
  if (!isRecord(value) || !exactKeys(value, [
    "displayName",
    "file",
    "fileId",
    "id",
    "isPrivateMethod",
    "kind",
    "qualifiedName",
    "stepCount",
  ]) || typeof value.id !== "string" || !nodeIdSchema.safeParse(value.id).success
    || typeof value.fileId !== "string" || !byId.has(value.fileId)
    || typeof value.displayName !== "string" || value.displayName.length === 0
    || typeof value.qualifiedName !== "string" || value.qualifiedName.length === 0
    || !INDEXED_KINDS.has(value.kind as NodeKind) || !safePath(value.file)
    || byId.get(value.fileId)?.path !== value.file
    || typeof value.isPrivateMethod !== "boolean" || value.stepCount !== null) {
    throw new TypeError("Python source symbol is invalid");
  }
  return value as unknown as CompactGraphSymbolEntry;
}

function bfs(
  topology: readonly PythonSourceTopologyFile[],
  seeds: readonly string[],
  depth: number,
  maxSelectedFiles: number,
): Omit<InitialPythonProjectionFiles, "extractionFiles"> | null {
  const byPath = new Map(topology.map((file) => [file.path, file]));
  const byId = new Map(topology.map((file) => [file.fileId, file]));
  const distance = new Map<string, number>();
  const queue: PythonSourceTopologyFile[] = [];
  for (const path of seeds) {
    const file = byPath.get(path);
    if (file === undefined) return null;
    distance.set(file.fileId, 0);
    queue.push(file);
  }
  for (let index = 0; index < queue.length; index += 1) {
    const file = queue[index]!;
    const nextDepth = distance.get(file.fileId)! + 1;
    if (nextDepth > depth) continue;
    for (const neighborId of file.neighbors) {
      if (distance.has(neighborId)) continue;
      const neighbor = byId.get(neighborId);
      if (neighbor === undefined) return null;
      distance.set(neighborId, nextDepth);
      if (distance.size > maxSelectedFiles) return null;
      queue.push(neighbor);
    }
  }
  const selected = topology.filter((file) => distance.has(file.fileId));
  const selectedIds = new Set(selected.map((file) => file.fileId));
  return {
    files: selected.map((file) => file.path),
    seeds: [...seeds],
    frontier: selected
      .filter((file) => file.uncertain || file.neighbors.some((neighbor) => !selectedIds.has(neighbor)))
      .map((file) => file.path),
  };
}

function normalizeSeeds(root: string, values: readonly string[]): string[] {
  const result = new Set<string>();
  for (const value of values) {
    if (!safePath(value) || isAbsolute(value)) throw new TypeError("Python source-index seed is invalid");
    let canonical: string;
    try {
      canonical = realpathSync.native(resolve(root, value));
    } catch {
      throw new TypeError("Python source-index seed is unavailable");
    }
    const normalized = relative(root, canonical).replaceAll("\\", "/");
    if (normalized !== value || normalized.startsWith("../")) {
      throw new TypeError("Python source-index seed escaped its workspace");
    }
    result.add(normalized);
  }
  return [...result].sort(compareBinaryStrings);
}

function canonicalDirectory(root: string): string {
  const canonical = realpathSync.native(root);
  if (!statSync(canonical).isDirectory()) throw new TypeError("Python source-index root is invalid");
  return canonical;
}

function existingDirectory(root: string): boolean {
  try {
    canonicalDirectory(root);
    return true;
  } catch {
    return false;
  }
}

function validateLimits(limits: PythonSourceIndexLimits): void {
  const pairs: Array<[number, number]> = [
    [limits.maxFiles, MAX_FILES],
    [limits.maxSourceBytes, MAX_SOURCE_BYTES],
    [limits.maxFileBytes, MAX_FILE_BYTES],
    [limits.maxSymbols, MAX_SYMBOLS],
    [limits.maxAdjacencyEntries, MAX_ADJACENCY],
  ];
  if (pairs.some(([value, maximum]) => !Number.isSafeInteger(value) || value <= 0 || value > maximum)) {
    throw new TypeError("Python source-index limits are invalid");
  }
}

function defaultLimits(): PythonSourceIndexLimits {
  return {
    maxFiles: MAX_FILES,
    maxSourceBytes: MAX_SOURCE_BYTES,
    maxFileBytes: MAX_FILE_BYTES,
    maxSymbols: MAX_SYMBOLS,
    maxAdjacencyEntries: MAX_ADJACENCY,
  };
}

function sortedPaths(values: unknown[], byPath: ReadonlyMap<string, unknown>): string[] {
  const result: string[] = [];
  const seen = new Set<string>();
  for (const value of values) {
    if (!safePath(value) || !byPath.has(value) || seen.has(value)) {
      throw new TypeError("Python source-index seeds are invalid");
    }
    seen.add(value);
    result.push(value);
  }
  return result.sort(compareBinaryStrings);
}

function pythonFileId(value: string): value is NodeId {
  const parsed = nodeIdSchema.safeParse(value);
  return parsed.success && value.startsWith("py:") && !value.includes("#");
}

function safePath(value: unknown): value is string {
  return typeof value === "string" && value.length > 0
    && Buffer.byteLength(value, "utf8") <= MAX_PATH_BYTES
    && value.endsWith(".py") && !value.startsWith("/") && !value.includes("\\") && !value.includes("\0")
    && value.split("/").every((part) => part.length > 0 && part !== "." && part !== "..");
}

function boundedInteger(value: unknown, minimum: number, maximum: number): number {
  if (!Number.isSafeInteger(value) || (value as number) < minimum || (value as number) > maximum) {
    throw new TypeError("Python source-index count is invalid");
  }
  return value as number;
}

function sameStrings(left: readonly string[], right: readonly string[]): boolean {
  return left.length === right.length && left.every((value, index) => value === right[index]);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function exactKeys(value: Record<string, unknown>, keys: readonly string[]): boolean {
  return Object.keys(value).sort(compareBinaryStrings).join("\0")
    === [...keys].sort(compareBinaryStrings).join("\0");
}

File: packages/extractor-python/python/source_index.py
#!/usr/bin/env python3
"""Bounded syntax-only Python topology and symbol scanner for progressive review.

The scanner never imports project code and retains at most one source AST.  Filesystem discovery
and import identity reuse the production extractor's helpers so selected-file extraction observes
the same module names.  Semantic extraction remains the authority after a slice is selected.
"""

from __future__ import annotations

import ast
import json
import os
import sys
import tokenize

from definitions import collect_nodes
from discovery import DiscoveredModule, discover_modules, module_aliases
from project import resolve_from_module_path


ROOT_PACKAGE = "__root__"


class BoundedFailure(Exception):
    pass


def main() -> None:
    if len(sys.argv) != 3:
        raise SystemExit("usage: source_index.py <root> <options-json>")
    root = os.path.abspath(sys.argv[1])
    options = read_options(sys.argv[2])
    discovered = discover_bounded(root, options)
    parsed = parse_bounded(discovered, options)
    parsed_paths = {item[0].file for item in parsed}
    # A changed file with invalid Python syntax has no semantic file node. Keep it in the exact PR
    # manifest, but exclude it from the graphable roots so preparation can land a manifest-only
    # empty side without inventing an ID that could perturb production package/module ordinals.
    seeds = sorted(set(options["seeds"]) & parsed_paths)

    file_ids, symbol_rows = allocate_ids(parsed, options["includeSymbols"])
    aliases = module_aliases(item[0] for item in parsed)
    module_by_path = {item[0].module_path: item[0] for item in parsed}
    adjacency: dict[str, set[str]] = {item[0].file: set() for item in parsed}
    adjacency_entries = 0
    for module, imports, _nodes, _size in parsed:
        for imported in imports:
            canonical = aliases.get(imported)
            target = module_by_path.get(canonical) if canonical else None
            if target is None or target.file == module.file:
                continue
            if target.file not in adjacency[module.file]:
                adjacency[module.file].add(target.file)
                adjacency[target.file].add(module.file)
                adjacency_entries += 2
                if adjacency_entries > options["maxAdjacencyEntries"]:
                    raise BoundedFailure("Python source index exceeded its adjacency limit")

    files = []
    for module, _imports, _nodes, _size in sorted(parsed, key=lambda item: item[0].file):
        files.append(
            {
                "path": module.file,
                "fileId": file_ids[module.file],
                "neighbors": sorted(file_ids[path] for path in adjacency[module.file]),
                "uncertain": False,
            }
        )
    symbol_rows.sort(key=lambda item: (item["displayName"], item["qualifiedName"], item["id"]))
    json.dump(
        {
            "seeds": seeds,
            "symbols": symbol_rows,
            "files": files,
            "sourceFileCount": len(parsed),
            "sourceBytes": sum(item[3] for item in parsed),
            "adjacencyEntries": adjacency_entries,
        },
        sys.stdout,
        separators=(",", ":"),
    )


def read_options(raw: str) -> dict:
    try:
        value = json.loads(raw)
    except json.JSONDecodeError as error:
        raise BoundedFailure("Python source-index options are invalid") from error
    if not isinstance(value, dict):
        raise BoundedFailure("Python source-index options are invalid")
    limits = (
        "maxFiles",
        "maxSourceBytes",
        "maxFileBytes",
        "maxSymbols",
        "maxAdjacencyEntries",
        "maxPathBytes",
    )
    for key in limits:
        if not isinstance(value.get(key), int) or value[key] <= 0:
            raise BoundedFailure("Python source-index limits are invalid")
    seeds = value.get("seeds")
    excludes = value.get("exclude")
    if (
        not isinstance(seeds, list)
        or not all(isinstance(item, str) and safe_path(item) for item in seeds)
        or not isinstance(excludes, list)
        or not all(isinstance(item, str) for item in excludes)
        or not isinstance(value.get("includeSymbols"), bool)
    ):
        raise BoundedFailure("Python source-index options are invalid")
    return {**value, "seeds": seeds, "exclude": excludes}


def discover_bounded(root: str, options: dict) -> list[DiscoveredModule]:
    result: list[DiscoveredModule] = []
    path_bytes = 0
    canonical_root = os.path.realpath(root)
    for module in discover_modules(root, (), options["exclude"]):
        if not safe_path(module.file):
            raise BoundedFailure("Python source index encountered an invalid source path")
        canonical_file = os.path.realpath(module.abs_path)
        try:
            confined = os.path.commonpath((canonical_root, canonical_file)) == canonical_root
        except ValueError:
            confined = False
        if not confined or canonical_file != os.path.join(canonical_root, *module.file.split("/")):
            raise BoundedFailure("Python source index encountered a source outside its exact workspace")
        result.append(module)
        if len(result) > options["maxFiles"]:
            raise BoundedFailure("Python source index exceeded its file limit")
        path_bytes += len(module.file.encode("utf-8"))
        if path_bytes > options["maxPathBytes"]:
            raise BoundedFailure("Python source index exceeded its path-byte limit")
    return result


def parse_bounded(
    discovered: list[DiscoveredModule],
    options: dict,
) -> list[tuple[DiscoveredModule, set[str], list[dict], int]]:
    result: list[tuple[DiscoveredModule, set[str], list[dict], int]] = []
    source_bytes = 0
    symbol_count = 0
    for module in discovered:
        size = os.path.getsize(module.abs_path)
        if size > options["maxFileBytes"]:
            raise BoundedFailure("Python source index encountered an oversized source file")
        source_bytes += size
        if source_bytes > options["maxSourceBytes"]:
            raise BoundedFailure("Python source index exceeded its source-byte limit")
        try:
            with tokenize.open(module.abs_path) as handle:
                tree = ast.parse(handle.read(), filename=module.file)
        except (OSError, SyntaxError, UnicodeError, ValueError):
            # The production analyzer omits the same unparseable source from its graph.
            continue
        imports = import_targets(module.module_path, tree)
        nodes = collect_nodes(tree) if options["includeSymbols"] else []
        if options["includeSymbols"]:
            symbol_count += len(nodes) + (0 if module.is_package else 1)
        if symbol_count > options["maxSymbols"]:
            raise BoundedFailure("Python source index exceeded its symbol limit")
        result.append((module, imports, nodes, size))
    return result


def import_targets(module_path: str, tree: ast.Module) -> set[str]:
    result: set[str] = set()
    for node in ast.walk(tree):
        if isinstance(node, ast.Import):
            result.update(alias.name for alias in node.names)
        elif isinstance(node, ast.ImportFrom):
            target = resolve_from_module_path(module_path, node)
            if target:
                result.add(target)
    return result


def allocate_ids(
    parsed: list[tuple[DiscoveredModule, set[str], list[dict], int]],
    include_symbols: bool,
) -> tuple[dict[str, str], list[dict]]:
    counts: dict[str, int] = {}

    def allocate(base: str) -> str:
        ordinal = counts.get(base, 0)
        counts[base] = ordinal + 1
        return base if ordinal == 0 else f"{base}~{ordinal}"

    prefixes = package_prefixes(item[0] for item in parsed)
    package_ids = {
        prefix: allocate(f"py:{prefix}")
        for prefix in sorted(prefixes, key=lambda value: (len(value.split(".")), value))
    }
    file_ids: dict[str, str] = {}
    symbols: list[dict] = []
    for module, _imports, nodes, _size in parsed:
        graph_path = graph_module_path(module)
        file_id = package_ids[graph_path] if module.is_package else allocate(f"py:{graph_path}")
        file_ids[module.file] = file_id
        if include_symbols and not module.is_package:
            symbols.append(
                symbol(file_id, file_id, graph_path.split(".")[-1], graph_path, "module", module.file)
            )
        for node in nodes:
            node_id = allocate(f"py:{graph_path}#{node['qualname']}")
            if include_symbols:
                symbols.append(
                    symbol(
                        node_id,
                        file_id,
                        node["name"],
                        node["qualname"],
                        node["kind"],
                        module.file,
                    )
                )
    return file_ids, symbols


def symbol(
    node_id: str,
    file_id: str,
    display_name: str,
    qualified_name: str,
    kind: str,
    file: str,
) -> dict:
    return {
        "id": node_id,
        "fileId": file_id,
        "displayName": display_name,
        "qualifiedName": qualified_name,
        "kind": kind,
        "file": file,
        "isPrivateMethod": kind == "method" and display_name.startswith("__"),
        "stepCount": None,
    }


def package_prefixes(modules) -> set[str]:
    prefixes: set[str] = set()
    count = 0
    for module in modules:
        count += 1
        graph_path = graph_module_path(module)
        segments = graph_path.split(".")
        for depth in range(1, len(segments)):
            prefixes.add(".".join(segments[:depth]))
        if module.is_package:
            prefixes.add(graph_path)
    if not prefixes and count:
        prefixes.add(ROOT_PACKAGE)
    return prefixes


def graph_module_path(module: DiscoveredModule) -> str:
    return module.module_path.removesuffix(".__init__") if module.is_package else module.module_path


def safe_path(value: str) -> bool:
    return (
        bool(value)
        and "\x00" not in value
        and "\\" not in value
        and not value.startswith("/")
        and value.endswith(".py")
        and all(part not in ("", ".", "..") for part in value.split("/"))
    )


if __name__ == "__main__":
    try:
        main()
    except BoundedFailure as error:
        sys.stderr.write(str(error))
        sys.exit(2)
